### ISSUE TITLE
Extracting interface causes circular dependencies error

### ISSUE THREAD

**Issue by ghost** [rubberduck-vba/Rubberduck#4737]
**Rubberduck version information**

Version 2.3.1.4457
OS: Microsoft Windows NT 10.0.17763.0, x64
Host Product: Microsoft Office x64
Host Version: 16.0.11231.20080
Host Executable: WINWORD.EXE

**Description**
I think this is a VBA problem but I'm reporting here for the sake of completeness.

My project has a couple of hand crafted Interfaces.  I'm now trying out the Extract interface command in Rubberduck to implement an interface for a new class.  Extracting the interface works fine.  However doing a compile immediately after gives the error message that private object modules cannot be used in public object modules.  If I set the Instancing of the newly created I<modulename> to PublicNotCreatable and then do a compile I get a 'Circular Dependencies Between Modules' error.

If I comment out the new implements statement the project compiles fine.

I have tested this on two new class modules and I get the same result of circular dependencies.

In both cases I have tried, continuing to Implement the interface does not change the error of circular references.

**Comment by SteveLaycock** [comment]
Yes.  The class is self instantiating so uses the Create/Self method of returning an instance of itself (the class has a predeclaredID).

**Comment by retailcoder** [comment]
Note that the pattern leaves the factory method and `Self` property getter on the class' default interface though.

**Comment by MDoerner** [comment]
This explains the problem. This is a bug indeed since the refactoring is very naive in how the interface gets extracted.

I am not completely sure how to deal with it though. Anyway, we should not extract an interface in a way that members still reference the original class.

**Comment by retailcoder** [comment]
Do you have an example that reproduces the problem? Tried this rather *minimal* example:

**Class1**

```vb
Public Property Get Self() As Class1
    Set Self = Me
End Property
```

Ran "extract interface", selected the `Self` getter, confirmed; I got this:

**Class1**

```vb
Implements IClass1

Public Property Get Self() As Class1
    Set Self = Me
End Property

Private Property Get IClass1_Self() As Class1
    Err.Raise 5 'TODO implement interface member
End Property
```

**IClass1**

```vb
Option Explicit

Public Property Get Self() As Class1
End Property
```

Project still compiles, no error, and VBA happily creates an instance of `Class1` without blinking.

How is the circular reference error triggered?

**Comment by MDoerner** [comment]
I see several options how to deal with members referencing the class type.

1. Do not extract such members at all. 
2. Only extract members where this only affects the return type and change the return type to the interface type. 
3. Extract all methods and change the class type to the interface type everywhere. 

Option 1 is clearly the most restrictive. 
Option 3 can lead to members that cannot be implemented because the necessary information is not exposed on the interface. 
Option 2 tries to strike a balance. It will always be possible to implement the interface members. However, they might be rather strange on the interface. An example of this is the `Self` factory method when used to construct new instances. (It makes perfect sense when used for casting purposes.) Arguably, we could take the position that the user should decide whether he wants the members with converted return type or not. (One can always delete them.) 

So, before starting to implement a fix we will have to decide on the approach to take.

**Comment by MDoerner** [comment]
@MathieuGuindon Did you try in Word?

**Comment by SteveLaycock** [comment]
Yes.  I'm working in Word VBA.

I tried the following simple class in my current project

```
Option Explicit

Implements IClass1

Public Function Create() As Class1

    With New Class1
    
        Set Create = .Self
        
    End With
    
End Function

Public Function Self() As Class1
        
    Set Self = Me
    
End Function

Public Function Test1() As Variant

End Function

Public Function Test2() As Variant

End Function

Private Function IClass1_Test1() As Variant
    Err.Raise 5 'TODO implement interface member
End Function

Private Function IClass1_Test2() As Variant
    Err.Raise 5 'TODO implement interface member
End Function
```
and it worked fine.  No Circular references.

I went back and removed an interface I added earlier today.  No circular references.  Then adding the interface back via rubber duck I again get circular references.

**Comment by SteveLaycock** [comment]
I've attached the class module and the resulting interface in case there is anything I've done in the class module that is causing the issue

[LoggerMain.zip](https://github.com/rubberduck-vba/Rubberduck/files/2788388/LoggerMain.zip)

**Comment by comintern** [comment]
Is the interface being extracted instanced as PublicNotCreateable?  That's the only way I can replicate this, and it's resolved by changing the interface to Private. Not that on my build, the instancing is Private when the interface is created.

**Comment by SteveLaycock** [comment]
I've tried compiling after changing the instancing to private and publicnotcreatable for the class and the interface.  I get the circular references error in both cases.

**Comment by comintern** [comment]
I just took a look at the code you posted, and the issue is related to `Public Enum ConcernType` and `Public Enum LocationType`.  You're exposing these on members of the interface, but they are declared on the implementing class.  All you should need to do in this case to make it compile is move the enumerations that are used by the interface to the interface.

I'm thinking we should probably include public `Enum`s in the refactoring.

**Comment by SteveLaycock** [comment]
Bingo!! Hopefully a useful contribution to Rubberduck.  I'm also really appreciative of the time  you guys to look at sorting this problem.
